### PR TITLE
rgw: d3n: fix d3n cached objects on disk lru eviction entering infini…

### PR DESCRIPTION
**Pull request by mkogan1**
…te loop

fixes: https://tracker.ceph.com/issues/52457

Signed-off-by: Mark Kogan <mkogan@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_d3n_datacache.cc [hunk 1]
```diff
@@ -235,27 +239,33 @@ void D3nDataCache::put(bufferlist& bl, unsigned int len, std::string& oid)
     _outstanding_write_size = outstanding_write_size;
   }
   ldout(cct, 20) << "D3nDataCache: Before eviction _free_data_cache_size:" << _free_data_cache_size << ", _outstanding_write_size:" << _outstanding_write_size << ", freed_size:" << freed_size << dendl;
-  while (len >= (_free_data_cache_size - _outstanding_write_size + freed_size)) {
-    ldout(cct, 20) << "D3nDataCache: enter eviction, r=" << r << dendl;
+  while (len > (_free_data_cache_size - _outstanding_write_size + freed_size)) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_d3n_datacache.cc:242]: is this the operative change (fixing the issue)?

File: src/rgw/rgw_d3n_datacache.cc [hunk 2]
```diff
@@ -161,9 +161,8 @@ void D3nDataCache::d3n_libaio_write_completion_cb(D3nCacheAioWriteRequest* c)
   { // update cache_map entries for new chunk in cache
     const std::lock_guard l(d3n_cache_lock);
     auto it = d3n_outstanding_write_list.find(c->oid);
-    if (it != d3n_outstanding_write_list.end()) {
+    if (it != d3n_outstanding_write_list.end())
       d3n_outstanding_write_list.erase(it);
-    }
```
Comment (cbodley) [inline src/rgw/rgw_d3n_datacache.cc:166]: please keep the {}s around single-line blocks, here and below (see [CodingStyle](https://github.com/ceph/ceph/blob/master/CodingStyle#L92-L103))

File: src/rgw/rgw_d3n_datacache.cc [hunk 3]
```diff
@@ -235,27 +239,33 @@ void D3nDataCache::put(bufferlist& bl, unsigned int len, std::string& oid)
...
       ldout(cct, 0) << "D3nDataCache: Warning: unknown cache eviction policy, defaulting to lru eviction" << dendl;
-      r = lru_eviction();
+      sr = lru_eviction();
     }
-    if (r < 0)
+    if (sr == 0) {
+      ldout(cct, 2) << "D3nDataCache: Warning: eviction was not able to free disk space, not writing to cache" << dendl;
+      auto it = d3n_outstanding_write_list.find(oid);
+      if (it != d3n_outstanding_write_list.end())
+        d3n_outstanding_write_list.erase(it);
```
Comment (cbodley) [inline src/rgw/rgw_d3n_datacache.cc:256]: can just call `erase(oid)` instead of find+erase

File: src/rgw/rgw_d3n_datacache.cc [hunk 4]
```diff
@@ -235,27 +239,33 @@ void D3nDataCache::put(bufferlist& bl, unsigned int len, std::string& oid)
...
     } else if (eviction_policy == _eviction_policy::RANDOM) {
-      r = random_eviction();
+      sr = random_eviction();
     } else {
       ldout(cct, 0) << "D3nDataCache: Warning: unknown cache eviction policy, defaulting to lru eviction" << dendl;
-      r = lru_eviction();
+      sr = lru_eviction();
     }
-    if (r < 0)
+    if (sr == 0) {
```
Comment (cbodley) [inline src/rgw/rgw_d3n_datacache.cc:252]: what would cause an eviction to return 0 here?

File: src/rgw/rgw_d3n_datacache.cc (hunk 1 above)
Comment (mkogan1) [reply src/rgw/rgw_d3n_datacache.cc:242]: yes, this `while (len >=` to `while (len >` change was necessary in order to be able to write to the cache an object that is exactly the size of the cache, ie if `rgw_d3n_l1_datacache_size` is 4 MiB and `s3cmd put` of a 8MiB object is performed the `while (len >=` kept trying to evict space even thou the cache is empty and can accomodate the 4MiB tail part from 8MiB object exactly

and also the change below is required:
https://github.com/ceph/ceph/blob/8ed8e8cd485c01dd9c5eb94f62dd5d1868881d3a/src/rgw/rgw_d3n_datacache.cc#L252-L258

where `sr` is the number of bytes evicted from the cache, if the cache cache is already empty an `0` will be returned to `sr` from the eviction.
before the fix the `while` loop at 
https://github.com/ceph/ceph/blob/8ed8e8cd485c01dd9c5eb94f62dd5d1868881d3a/src/rgw/rgw_d3n_datacache.cc#L242
kept spinning indefinitely as the check for the condition that there is nothing to evict was incorrect

**Review by cbodley — APPROVED** [review]

**Pull request merged**